src/layout.cpp

```diff
@@ -0,0 +1,408 @@
+          auto edge = std::make_pair(nodes[i].id, nodes[i].succ[j]);
+          deletedEdges.push_back(edge);
+
+          nodes[i].succ.erase(nodes[i].succ.begin() + j);
+        }
+      }
+    }
+  }
+
+  std::sort(nodes.begin(), nodes.end(), comparatorId);
+    
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].pred.clear();
+  }
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      nodes[nodes[i].succ[j]].pred.push_back(i);
+    }
+  }
+}
+
+void algorithmASAP(
+    std::vector<TreeNode> &nodes,
+    std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges,
+    std::vector<int> &lensLayer) {
+  std::vector<int> removedNodes(nodes.size(), -1);
+
+    std::vector<int> degInNodes;
+    degInNodes.reserve(nodes.size());
+    for (const TreeNode &node: nodes) {
+        degInNodes.push_back(node.pred.size());
+    }
+
+  int nodesInLayer = 0;
+  int removedCount = 0;
+  for (int i = 0; removedCount < nodes.size(); i++) {
+    for (int j = 0; j < nodes.size(); j++) {
+      if (degInNodes[j] == 0 && removedNodes[j] == -1) {
+        removedNodes[j] = 0;
+        removedCount++;
+      }
+    }
+
+    for (int j = 0; j < nodes.size(); j++) {
+      if (removedNodes[j] == 0) {
+        nodes[j].layer = i;
+        nodes[j].number = nodesInLayer;
+        nodesInLayer++;
+
+        for (int k = 0; k < nodes[j].succ.size(); k++) {
+          degInNodes[nodes[j].succ[k]]--;
+        }
+        removedNodes[j] = 1;
+      }
+    }
+    lensLayer.push_back(nodesInLayer);
+    nodesInLayer = 0;
+  }
+
+  for (int i = 0; i < deletedEdges.size(); i++) {
+    nodes[deletedEdges[i].first].succ.push_back(deletedEdges[i].second);
+    nodes[deletedEdges[i].second].pred.push_back(deletedEdges[i].first);
+  }
+}
+
+void addLineSegment(std::vector<TreeNode> &nodes, std::vector<int> &lensLayer,
+                    int idStart, int idEnd, int order) {
+  int idPrevSucc = nodes[idStart].succ[idEnd];
+  nodes[idStart].succ[idEnd] = nodes.size();
+  for (int i = nodes[idStart].layer + order;
+       (i < nodes[idPrevSucc].layer && order == 1) ||
+       (i > nodes[idPrevSucc].layer && order == -1);
+       i = i + order) {
+    TreeNode dummy = {};
+    dummy.isDummy = true;
+    dummy.id = static_cast<int>(nodes.size());
+    dummy.layer = i;
+    dummy.number = lensLayer[i];
+    lensLayer[i]++;
+    if (i + order == nodes[idPrevSucc].layer) {
+      dummy.succ.push_back(idPrevSucc);
+    } else {
+      dummy.succ.push_back(dummy.id + 1);
+    }
+    nodes.push_back(dummy);
+  }
+}
+
+bool addDummyNodes(std::vector<TreeNode> &nodes, std::vector<int> &lensLayer) {
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      if ((nodes[nodes[i].succ[j]].layer - nodes[i].layer > 1) || (nodes[nodes[i].succ[j]].layer - nodes[i].layer < -1)) {
+        int order = -1;
+        if (nodes[nodes[i].succ[j]].layer - nodes[i].layer > 1) {
```

Comment:
Please, stop copying nested indexes and use proper named references.
```suggestion
  for (const TreeNode &node: nodes) {
    for (TreeNode::nodeId succId: node.succ) {
      if ((nodes[succId].layer - node.layer > 1) || (nodes[succId].layer - node.layer < -1)) {
        int order = -1;
        if (nodes[succId].layer - node.layer > 1) {
```
Code duplication, btw